How many late days are you using for this assignment? 

0

In [106]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [107]:

# Load the given datasets
df_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
df_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")


# Print the data
df_categorical.head()

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


In [108]:
df_numerical.head()

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN


In [109]:
df_categorical.dtypes



unique_id      int64
al           float64
su           float64
rbc           object
pc            object
pcc           object
ba            object
htn           object
dm            object
cad           object
appet         object
pe            object
ane           object
Target        object
dtype: object

In [110]:
df_numerical.dtypes

unique_id      int64
age          float64
bp           float64
bgr          float64
bu           float64
sc           float64
sod          float64
pot          float64
hemo         float64
pcv          float64
wbcc         float64
rbcc         float64
dtype: object

# Part 2: Analyze the Dataset

In [111]:
features = df_categorical.columns.tolist()
features

['unique_id',
 'al',
 'su',
 'rbc',
 'pc',
 'pcc',
 'ba',
 'htn',
 'dm',
 'cad',
 'appet',
 'pe',
 'ane',
 'Target']

In [112]:
features_numerical = df_numerical.columns.tolist()
features_numerical

['unique_id',
 'age',
 'bp',
 'bgr',
 'bu',
 'sc',
 'sod',
 'pot',
 'hemo',
 'pcv',
 'wbcc',
 'rbcc']

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data is chronic kidney disease aka CKD and it contains  of two  files: categorical and numerical data. The dataset is used to predict the chronic kidney disease and it can be gathered from the hospital over a span of approximately two months. The dataset has 24 features which contains categorical & numerical, and also contains the medical attributes. The label has a role of "Target" and variable name "class". It is to determine if the patient has chronic kidney diseases or not. 

Examples: 

Categorical (Binary, Categorical - discrete values)
1. rbc - red blood cells	
2. su - sugar 
3. pcc - ps cell clumps 

Numerical (Integer, Continuous - continuous values)
1. age - Age
2. bgr - blood glucose random 
3. sc - serum creatinine	

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [113]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
df_check_duplicated = df_numerical.duplicated().sum()
print("Numerical:", df_check_duplicated)

# Drop these duplicate rows
df_numerical.drop_duplicates(inplace=True)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
df_check_duplicated = df_categorical.duplicated().sum()
print("Categorical:",df_check_duplicated)

df_categorical.drop_duplicates(inplace=True)


Numerical: 30
Categorical: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [114]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df_combined = pd.merge(df_numerical, df_categorical, on='unique_id')
#Print the combined dataset
print(df_combined)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

In [115]:
df_combined.head()

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [116]:
nan_count = df_combined.isnull().sum() # To help check if the df contains the missing value 
nan_count

unique_id      0
age            8
bp            12
bgr           42
bu            19
sc            17
sod           85
pot           86
hemo          51
pcv           71
wbcc         104
rbcc         129
al            44
su            47
rbc          149
pc            63
pcc            3
ba             3
htn            2
dm             2
cad            2
appet          1
pe             1
ane            1
Target         0
dtype: int64

In [117]:
# Calculate the percentage of rows that contain atleast one missing value
df_percentage_missing = df_combined.isnull().any(axis=1).sum()/len(df_combined)*100


# Print %
print(f"{df_percentage_missing:.2f}%")

# Drop these rows from the dataset
df_combined.dropna(inplace=True)

# Print the Dataset
print(df_combined)

60.77%
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc         pcc          ba  htn   dm  cad app

## Part 3.4: Sort the dataset according to the Labels

In [118]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df_combined = df_combined.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
print(df_combined)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc         pcc          ba  htn   dm  ca

In [119]:
df_combined.head(100) # for better formatting of the output

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,328705,39.0,70.0,131.0,46.0,0.6,145.0,5.0,15.6,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
96,268609,56.0,80.0,139.0,15.0,1.2,135.0,5.0,14.8,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
97,524004,30.0,80.0,82.0,42.0,0.7,146.0,5.0,14.9,45.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
98,418316,62.0,80.0,78.0,45.0,0.6,138.0,3.5,16.1,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [120]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

df_combined = pd.get_dummies(df_combined, columns=categorical_columns, drop_first=True)
# Print the dataset
print(df_combined)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

In [121]:
df_combined.head()

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc_normal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_poor,pe_yes,ane_yes,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,False,True,True,False,False,False,False,False,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,False,True,False,True,True,True,True,False,False,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,False,True,False,True,True,False,True,False,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,False,False,False,True,True,True,True,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,True,False,False,False,False,False,False,False,False,False


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Having a single column for the binary variable would mean that it will give a similar results (especially if we use encoding) cause the output can be either 0 or 1. If the categorical variable take up more than 2 values, then it can mislead the model to think that there is an order or that one value can be greater than other. The encoding avoid the issues by separating the columns and make each category independently or can make sure all values are being treated. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [122]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for i in numerical_columns:
    mean = df_combined[i].mean()
    std = df_combined[i].std()
    lower_bound = mean - (3*std)
    upper_bound = mean + (3*std)
    
    df_combined = df_combined[(df_combined[i] > lower_bound) & (df_combined[i] < upper_bound)]
    
# Print the dataset
print(df_combined)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [123]:
# Normalize the all Numerical Attributes in the dataset.
df_combined[numerical_columns] = (df_combined[numerical_columns] - df_combined[numerical_columns].min()) / (df_combined[numerical_columns].max() - df_combined[numerical_columns].min())

# Print the dataset
print(df_combined)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
3       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
6       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
148     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
149     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
150     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
151     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
152     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv  ...  pc_normal  pcc_present  ba_present  \
0    0.206897  0.059406  0.000000

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: I believe unique_id is not useful for the modeling and predications because it is a identifer that is being assigned to patients and does not contribute on how it can predict the CKD. The model is not trying to find patterns with the IDs. 

unique_id can be useful when we are trying to merge the datasets and match records correctly 

In [124]:
#Remove that column
df_combined.drop('unique_id', axis=1, inplace=True)
# Print the dataset
print(df_combined)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \
0    0.059406  0.000000  0.653226  ...      F

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [125]:
# Export the dataframe to a new csv file
df_combined.to_csv("chronic_kidney_disease_cleaned.csv", index=False)
# Export the dataframe to a new json file
df_combined.to_json("chronic_kidney_disease_cleaned.json", orient='records')

print(df_combined)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \
0    0.059406  0.000000  0.653226  ...      F

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

| Age       | BP   | BGR               | BU                | SC               | SOD               | POT               | Hemo             | PCV              | WBCC             | RBCC            | AL               | SU       | RBC Normal | PC Normal | PCC Present | BA Present | HTN Yes | DM Yes | CAD Yes | Appetite Poor | PE Yes | Ane Yes | Target Not CKD |
|-----------|------|-------------------|-------------------|-------------------|-------------------|-------------------|-------------------|-------------------|-------------------|------------------|-------------------|----------|------------|-----------|-------------|------------|---------|--------|---------|----------------|--------|--------|----------------|
| 0.20270   | 0.75 | 0.158798283261803 | 0.196078431372549 | 0.160493827160494 | 0.166666666666667 | 0.206896551724138 | 0.059405940594059 | 0.0               | 0.653225806451613 | 0.257142857142857 | 4.0               | 0.0      | TRUE       | FALSE     | TRUE        | TRUE       | FALSE   | FALSE  | FALSE   | FALSE          | TRUE   | FALSE  |                |
| 0.90541   | 1.0  | 0.965665236051502 | 0.522875816993464 | 0.641975308641975 | 0.666666666666667 | 0.0               | 0.148514851485148 | 0.225806451612903 | 0.217741935483871 | 0.057142857142857 | 3.0               | 2.0      | FALSE      | FALSE     | TRUE        | FALSE      | TRUE    | TRUE   | TRUE    | TRUE           | FALSE  | FALSE  |                |
| 0.74324   | 0.5  | 0.442060085836910 | 0.901960784313726 | 0.432098765432099 | 0.5               | 0.793103448275862 | 0.0               | 0.032258064516129 | 0.395161290322581 | 0.057142857142857 | 2.0               | 0.0      | FALSE      | FALSE     | FALSE       | FALSE      | TRUE    | TRUE   | TRUE    | TRUE           | TRUE   | TRUE   | FALSE          |
| 0.72973   | 0.75 | 0.150214592274678 | 0.281045751633987 | 0.234567901234568 | 0.533333333333333 | 0.793103448275862 | 0.336633663366337 | 0.322580645161290 | 0.5               | 0.314285714285714 | 2.0               | 0.0      | FALSE      | TRUE      | FALSE       | FALSE      | FALSE   | FALSE  | FALSE   | FALSE          | FALSE  | FALSE  | FALSE          |
| 0.54054   | 0.0  | 0.399141630901288 | 0.535947712418301 | 0.358024691358025 | 0.7               | 0.379310344827586 | 0.207920792079208 | 0.161290322580645 | 0.830645161290323 | 0.057142857142857 | 1.0               | 0.0      | TRUE       | TRUE      | FALSE       | FALSE      | TRUE    | TRUE   | FALSE   | FALSE          | FALSE  | FALSE  | FALSE          |
| 0.67568   | 0.75 | 0.253218884120172 | 0.633986928104575 | 0.777777777777778 | 0.366666666666667 | 0.655172413793103 | 0.138613861386139 | 0.193548387096774 | 0.169354838709677 | 0.114285714285714 | 2.0               | 0.0      | FALSE      | FALSE     | FALSE       | FALSE      | TRUE    | FALSE  | FALSE   | FALSE          | FALSE  | FALSE  | FALSE          |
| 0.71622   | 0.5  | 1.0               | 0.163398692810458 | 0.111111111111111 | 0.066666666666667 | 0.206896551724138 | 0.267326732673267 | 0.387096774193548 | 0.532258064516129 | 0.371428571428571 | 1.0               | 0.0      | FALSE      | TRUE      | FALSE       | FALSE      | TRUE    | FALSE  | FALSE   | FALSE          | FALSE  | FALSE  | FALSE          |
| 0.72973   | 0.0  | 0.935622317596567 | 0.169934640522876 | 0.160493827160494 | 0.333333333333333 | 0.034482758620689 | 0.019801980198020 | 0.064516129032258 | 0.879032258064516 | 0.0               | 3.0               | 1.0      | TRUE       | FALSE     | TRUE        | FALSE      | TRUE    | FALSE  | FALSE   | FALSE          | TRUE   | FALSE  | FALSE          |
| 0.0       | 0.0  | 0.103004291845494 | 0.372549019607843 | 0.074074074074074 | 0.5               | 0.689655172413793 | 0.217821782178218 | 0.225806451612903 | 1.0               | 0.514285714285714 | 4.0               | 0.0      | FALSE      | FALSE     | TRUE        | FALSE      | FALSE   | FALSE  | FALSE   | FALSE          | TRUE   | FALSE  | FALSE          |
| 0.87838   | 0.0  | 0.206008583690987 | 0.751633986928105 | 0.604938271604938 | 0.533333333333333 | 0.689655172413793 | 0.366336633663366 | 0.387096774193548 | 0.879032258064516 | 0.371428571428571 | 4.0               | 0.0      | TRUE       | TRUE      | FALSE       | FALSE      | TRUE    | TRUE   | FALSE   | TRUE           | TRUE   | FALSE  | FALSE          |
| 0.85135   | 0.25 | 0.618025751072961 | 0.562091503267974 | 0.728395061728395 | 0.0               | 0.344827586206897 | 0.168316831683168 | 0.161290322580645 | 0.580645161290323 | 0.085714285714286 | 4.0               | 3.0      | TRUE       | FALSE     | TRUE        | TRUE       | TRUE    | TRUE   | FALSE   | TRUE           | TRUE   | FALSE  | FALSE          |
| 0.87838   | 0.25 | 0.639484978540773 | 0.470588235294118 | 0.395061728395062 | 0.433333333333333 | 0.517241379310345 | 0.267326732673267 | 0.322580645161290 | 0.104838709677419 | 0.171428571428571 | 3.0               | 0.0      | TRUE       | FALSE     | TRUE        | TRUE       | TRUE    | TRUE   | FALSE   | FALSE          | FALSE  | FALSE  | FALSE          |
| 0.78378   | 0.0  | 0.725321888412017 | 0.313725490196078 | 0.481481481481482 | 0.566666666666667 | 0.862068965517242 | 0.178217821782178 | 0.193548387096774 | 0.258064516129032 | 0.114285714285714 | 4.0               | 1.0      | FALSE      | FALSE     | TRUE        | TRUE       | TRUE    | FALSE  | TRUE    | TRUE           | TRUE   | FALSE  | FALSE          |
| 0.66216   | 0.5  | 0.618025751072961 | 0.411764705882353 | 0.432098765432099 | 0.566666666666667 | 0.689655172413793 | 0.316831683168317 | 0.354838709677419 | 0.25              | 0.200000000000000 | 3.0               | 1.0      | TRUE       | FALSE     | TRUE        | TRUE       | TRUE    | TRUE   | FALSE   | FALSE          | FALSE  | FALSE  | FALSE          |
| 0.77027   | 1.0  | 0.901287553648069 | 0.163398692810458 | 0.345679012345679 | 0.766666666666667 | 0.206896551724138 | 0.524752475247525 | 0.548387096774194 | 0.443548387096774 | 0.342857142857143 | 2.0               | 2.0      | TRUE       | TRUE      | FALSE       | TRUE       | TRUE    | FALSE  | FALSE   | FALSE          | FALSE  | FALSE  | FALSE          |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

We can verify the dataset by doing the analysis process. We should make sure that we have the original dataset and converted or cleaned dataset save at the different file so we don't lose our process

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [126]:
# Code to rename all the columns in the dataset

df_combined.rename(columns={
    'age_in_years': 'Age', 
    'bp_mmHg': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell clumps',
    'ba': 'Bacteria',
    'bgr_mg_dl': 'Blood Glucose Random',
    'bu_mg_dl': 'Blood Urea',
    'sc_mg_dl': 'Serum Creatinine',
    'sod_mEq_L': 'Sodium',
    'pot_mEq_L': 'Potassium',
    'hemo_gms': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wc_cells_cumm': 'White Blood Cell Count',
    'rc_millions_cmm': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'Class'
}, inplace=True)

print(df_combined)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo  Packed Cell Volume      wbcc  ...  pc_normal  pcc_present  \
0    0.059406            0.000000  0.653226  ..

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [127]:
# Converted SQL to Pandas code
result = df_combined.groupby('Target_notckd').size().reset_index(name='count')

print(result)

   Target_notckd  count
0          False     22
1           True    111


*Describe what the above code does here*

The code shows how many times each unique value of Target_notckd appears in the dataframe. The output shows that there are 22 count of the target having Chronic Kidney Diseases and 111 count that the target does not have chronic kidney diseases 